src/bosh-director/lib/bosh/director/job_queue.rb

```diff
@@ -39,7 +42,14 @@ def create_task(username, type, description, deployment, context_id)
       end
 
       task.output = log_dir
-      task.save
+      Config.db.transaction(
+        wait: 1.seconds,
+        attempts: 5,
+        retry_on: [Sequel::DatabaseConnectionError, Sequel::DatabaseDisconnectError],
+      ) do
+        task.save
```

Comment:
The transaction block here is not idempotent. When "save" is called on sequel models, the library checks to see if the model is `@new`, and if so, [performs an insert and sets `@new` to `false`](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/model/base.rb#L1787-L1793), if the model is NOT new, it performs an update.

If the transaction here fails after the save has happened, during the transaction commit, the `task` model will already have cleared the `@new` attribute, so trying to call `task.save` a second time during the retry will try to `update` the record in the database, but it doesn't currently exist due to the first transaction being rolled back and will throw an error.

Reply:
going over this again, I think I was a bit overzealous adding the retry here as well. Since this are of the code wasn't even involved in the problem that this tries to solve, I just removed this part of the change.